core/src/main/java/org/apache/calcite/util/Sources.java

```diff
@@ -75,7 +94,68 @@ private static boolean isFile(Source source) {
     return source.protocol().equals("file");
   }
 
-  /** Implementation of {@link Source}. */
+  /**
+   * Adapter for {@link CharSource}
+   */
+  private static class GuavaCharSource implements Source {
+    private final CharSource charSource;
+
+    private GuavaCharSource(CharSource charSource) {
+      this.charSource = Objects.requireNonNull(charSource, "charSource");
+    }
+
+    private UnsupportedOperationException unsupported() {
+      return new UnsupportedOperationException(String.format(Locale.ROOT,
+          "Invalid operation for '%s' protocol", protocol()));
+    }
+
+    @Override public URL url() {
+      throw unsupported();
+    }
+
+    @Override public File file() {
+      throw unsupported();
```

Comment:
Should it be added as a default implementation to `Source` ?

Reply:
You're suggesting changing Source interface like:

```java
interface Source {
   default File file() {
      throw new UnsupportedOperationException();
   }
}
```

Reply:
Ah, there are just two implementations. Then it does not matter much.
I just thought we could pull `throw new UnsupportedOperationException` to the interface default methods.